synapse/storage/prepare_database.py

```diff
@@ -108,9 +108,15 @@ def prepare_database(
         # so we start one before running anything. This ensures that any upgrades
         # are either applied completely, or not at all.
         #
-        # (psycopg2 automatically starts a transaction as soon as we run any statements
-        # at all, so this is redundant but harmless there.)
-        cur.execute("BEGIN TRANSACTION")
+        # psycopg2 does not automatically start transactions when in autocommit mode.
+        # While it is technically harmless to nest transactions in postgres, doing so
+        # results in a warning in Postgres' logs per query. And we'd rather like to
+        # avoid doing that.
+        if isinstance(database_engine, Sqlite3Engine) or (
+            isinstance(database_engine, PostgresEngine)
+            and db_conn.autocommit
+        ):
```

Comment:
Is `db_conn.autocmmit` always false for sqlite? Can we simplify this to:

```suggestion
        if db_conn.autocommit:
```

Note that attempting to set auto-commit for sqlite seems to be a no-op:

https://github.com/matrix-org/synapse/blob/8c94dd3a277d4e11192f98a9ca32cb6638606b66/synapse/storage/engines/sqlite.py#L114-L119

Reply:
From playing around in the manhole, it sadly appears that `autocommit` isn't a valid attribute in the case of SQLite:

```
>>> hs.get_datastores().main.db_pool.runWithConnection(lambda conn: print("ac:", conn.autocommit))
<Deferred #0>
Deferred #0 failed: "'sqlite3.Connection' object has no attribute 'autocommit'"
```

Reply:
https://docs.python.org/3/library/sqlite3.html?highlight=sqlite3#transaction-control

> The sqlite3 module does not adhere to the transaction handling recommended by [PEP 249](https://peps.python.org/pep-0249/).

🫠🫠

Reply:
That's annoying. I secretly want to burn all `isinstance(foo, FooEngine)`....but not a big deal. 👍

Reply:
It looks like sqlite can be in an autocommit mode or an implicit transaction mode, depending on the value of isolation_level.

It'd be nice if we could abstract these details away into the DB engines somehow. The ideal would be something like 

```
if db_engine.will_autocommit(db_conn):
    cur.execute("BEGIN TRANSACTION")
```

but I don't know if that's going to open up an even bigger can of worms. Maybe see what the rest of the team think?

Reply:
The abstraction sounds like a good idea. I'd not block the PR on it though.